File: system-test/typescript/src/testkit/dsl/core/usecase/external/erp/ErpDsl.ts
import type { ErpDriver } from '../../../../../driver/port/external/erp/erp-driver.js';
import type { UseCaseContext } from '../../../use-case-context.js';
import { ReturnsProduct } from './usecases/ReturnsProduct.js';
import { ReturnsPromotion } from './usecases/ReturnsPromotion.js';
import { GetProduct } from './usecases/GetProduct.js';
import { GoToErp } from './usecases/GoToErp.js';

export class ErpDsl {
  constructor(
    private readonly driver: ErpDriver,
    private readonly context: UseCaseContext,
  ) {}

  goToErp(): GoToErp {
    return new GoToErp(this.driver, this.context);
  }

  returnsProduct(): ReturnsProduct {
    return new ReturnsProduct(this.driver, this.context);
  }

  returnsPromotion(): ReturnsPromotion {
    return new ReturnsPromotion(this.driver, this.context);
  }

  getProduct(): GetProduct {
    return new GetProduct(this.driver, this.context);
  }

  async close(): Promise<void> {
    await this.driver.close();
  }
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/external/erp/ErpDsl.java
package com.mycompany.myshop.testkit.dsl.core.usecase.external.erp;

import com.mycompany.myshop.testkit.driver.port.external.erp.ErpDriver;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.erp.usecases.GetProduct;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.erp.usecases.GoToErp;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.erp.usecases.ReturnsProduct;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.erp.usecases.ReturnsPromotion;
import com.mycompany.myshop.testkit.common.Closer;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;

public class ErpDsl implements AutoCloseable {
    protected final ErpDriver driver;
    protected final UseCaseContext context;

    public ErpDsl(ErpDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToErp goToErp() {
        return new GoToErp(driver, context);
    }

    public ReturnsProduct returnsProduct() {
        return new ReturnsProduct(driver, context);
    }

    public GetProduct getProduct() {
        return new GetProduct(driver, context);
    }

    public ReturnsPromotion returnsPromotion() {
        return new ReturnsPromotion(driver, context);
    }
}
